dali/kernels/audio/mel_scale/mel_filter_bank_gpu.cu

```diff
@@ -0,0 +1,306 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <vector>
+#include <memory>
+#include "dali/kernels/audio/mel_scale/mel_filter_bank_gpu.h"
+#include "dali/core/tensor_shape_print.h"
+
+namespace dali {
+namespace kernels {
+namespace audio {
+
+const int kBlockDim2 = 32;
+const int kBlockDim1 = 1024;
+
+template <typename T>
+struct BlockDesc {
+  const T *in_frame;
+  T *out_frame;
+  union {
+    struct {  // outer-dim fft
+      int64_t start_window;
+      int64_t frame_nwindows;
+    };
+    struct {  // inner-dim fft
+      int64_t block_start;
+      int64_t out_frame_size;
+    };
+  };
+};
+
+template <typename T>
+__device__ T calcMel(const T* in_frame, int mel_bin,
+                     const T *weights_down, const int *interval_ends,
+                     int fft_stride, int fft_shift,
+                     T norm_factor) {
+  T out = 0;
+  int fftbin = interval_ends[mel_bin];
+  int fftbin_end = interval_ends[mel_bin + 1];
+  const T *in =  in_frame + fftbin * fft_stride + fft_shift;
+  for (; fftbin < fftbin_end; ++fftbin, in += fft_stride) {
+    auto weight_up = T(1) - weights_down[fftbin];
+    weight_up *= norm_factor;
+    out += *in * weight_up;
+  }
+  fftbin = interval_ends[mel_bin + 1];
```

Comment:
I think this line is a no-op - either the bin belongs to one slope or the other.

Reply:
done